Add support for fingerprint alg

bazsi opened the pull request:
This branch allows you to specify the fingerprint-alg() to be used for validating key fingerprints in trusted-keys()

Fixes: https://github.com/syslog-ng/syslog-ng/issues/4976

alltilla reviewed (commented):
The last commit lacks the signoff part. I don't think it is a good idea if I write it for you and force push, so please add it yourself. Thanks!

The code looks good to me. We need a newsfile entry, but I can make it for you.

bazsi committed:
news: added news file

Signed-off-by: Balazs Scheidler <balazs.scheidler@axoflow.com>

bazsi commented:
I've rebased, fixed the signoff and added the news entry. btw, @MrAnno this PR now allows to trust specific keys, even if their X.509 validation fails. 

The code did not support it thus far, but I think this has always been the intention. Currently, trusted-keys() will only trust keys that are otherwise trusted, which does not make much sense to me.

The documentation is not that clear on this topic either (@fekete-robert).

bazsi commented:
I've dropped the change in trusted-keys() validation, as we agreed with @MrAnno that it should be introduced with a new option (trusted-fingerprints() or something like that).

By dropping this patch, this PR became very simple, _if_ there's trusted-keys() based validation in a config, the key fingerprint is saved into ${.tls.x509_fp} along with the CA and similar parameters.

With that this is mergable.

bazsi committed:
tls-context: move tls_session_set_trusted_fingerprints/dns

These functions belong to TLSContext, they just happened to remain in
TLSSession when the two were split.

Signed-off-by: Balazs Scheidler <balazs.scheidler@axoflow.com>

MrAnno commented:
From #136:

> I was not completely right here: if the fingerprint alg is chosen well, it is suitable to replace the CA-based verification, and that can be a perfectly valid use case.

> With this new implementation, we should not default to SHA1, and we should consider not letting users set a weak hash function.
To avoid any confusion, I'd remove the old option and introduce a new one, for example:

```
trusted-fingerprints(sha256, "aa:bb:cc", ...)
```

I can implement the missing parts next week.

bazsi committed:
tls-session: add fingerprint_alg parameter to tls_x509_digest()

Signed-off-by: Balazs Scheidler <balazs.scheidler@axoflow.com>

bazsi committed:
tls-session: add trusted-fingerprints() validation

trusted-fingerprints() is the new name of the now deprecated
trunsted-keys().  The primary difference is that trusted-fingerprints() will
trust the key even if CA based validation fails, whereas trusted-keys() is
checking and continues to check if the CA validation succeeds as well.

Signed-off-by: Balazs Scheidler <balazs.scheidler@axoflow.com>

bazsi committed:
tls-session: print validation errors everywhere

This also has a slight refactor of _log_certificate_validation_progress()

Signed-off-by: Balazs Scheidler <balazs.scheidler@axoflow.com>

bazsi commented:
I rebased and finished off the branch to do what was discussed in comments above.

bazsi commented:
This should be back on the review queue

bazsi commented on lib/transport/tls-session.c:
```diff
@@ -297,6 +366,13 @@ tls_session_verify_callback(int ok, X509_STORE_CTX *ctx)
     }
+  else if (self->cert_trusted_by_fingerprint)
+    {
+      msg_debug("Certificate invalid, but accepting due to being present on trusted-fingerprints()",
```
Having a matching fingerprint is a very narrow and specific case, matching against the public key of the x.509 keypair.

I don't think we should do any additional validation in this case. This is the same as the ssh known_hosts authentication, where we _only_ have the fingerprint, this is very similar.
